ndonnx/extensions.py

```diff
@@ -0,0 +1,437 @@
+
+    Returns
+    -------
+    out: Array
+        Array of booleans indicating whether each element of ``x`` is in ``items``.
+    """
+    return ndx.Array._from_tyarray(x._tyarray.isin(items))
+
+
+# TODO: Bad naming: It is obvious from the type hints that this mapping is "static"
+def static_map(
+    x: ndx.Array,
+    /,
+    mapping: Mapping[Key, Value],
+    default: Value | None = None,
+) -> ndx.Array:
+    """Map values in ``x`` based on the static ``mapping``.
+
+    Parameters
+    ----------
+    x: Array
+        The Array whose values will be mapped.
+    mapping: Mapping[Key, Value]
+        A mapping from keys to values. The keys must be of the same type as the values in ``x``.
+    default: Value, optional
+        The default value to use when a key is not found in the mapping.
+
+    Returns
+    -------
+    out: Array
+        A new Array with the values mapped according to the mapping.
+    """
+    values = np.array(list(mapping.values()))
+
+    if values.dtype.kind in ("O", "U"):
+        values = values.astype(str)
+
+    if default is None:
+        # Default values do not follow the ONNX standard (which is ok!)
+        if np.issubdtype(values.dtype, np.floating):
+            default = 0.0  # type: ignore
+        elif np.issubdtype(values.dtype, np.integer):
+            default = 0  # type: ignore
+        elif values.dtype.kind == "U":
+            default = "MISSING"  # type: ignore
+        elif values.dtype == bool:
+            # to be in line with the numerical 0
+            default = False  # type: ignore
+    if default is None:
+        raise TypeError(
+            f"failed to infer default value for mapping values of data type `{values.dtype}`"
+        )
+
+    return ndx.Array._from_tyarray(x._tyarray.apply_mapping(mapping, default))
+
+
+def fill_null(x: ndx.Array, /, value: ndx.Array | Scalar) -> ndx.Array:
+    """Returns a new ``Array`` with the null values filled with the given value.
+
+    Parameters
+    ----------
+    x: Array
+        The Array to fill the null values of.
+    value: Array | Scalar
+        The value to fill the null values with.
+
+    Returns
+    -------
+    out: Array
+        A new Array with the null values filled with the given value.
+    """
+    value_ = ndx.asarray(value)._tyarray
+    if isinstance(value_.dtype, ndx.Nullable):
+        raise ValueError("'fill_null' expects a none-nullable fill value data type")
+    xty = x._tyarray
+    if isinstance(xty, tydx.masked_onnx.TyMaArray):
+        result_type = ndx.result_type(xty.data.dtype, value_.dtype)
+        if xty.mask is None:
+            res: tydx.TyArrayBase = xty.data
+        else:
+            res = tydx.funcs.where(xty.mask, value_, xty.data)
+    elif isinstance(xty, tydx.onnx.TyArray):
+        result_type = ndx.result_type(xty.dtype, value_.dtype)
+        res = xty.astype(result_type)
+    else:
+        raise TypeError("'fill_null' is only implemented for built-in types")
```

Comment:
I believe this is a regression (see https://github.com/Quantco/ndonnx/blob/main/ndonnx/additional/_additional.py#L115, any dtype should be able to implement this). Can you offload it to the typed array in the style of the other functions?

Reply:
There is no standardized notion of what `null` is meant to be for a general data type and I'd like to keep it that way. In these terms it does not make sense to have this function be implemented by custom data types.